presto-main/src/main/java/com/facebook/presto/server/remotetask/ContinuousTaskStatusFetcher.java

```diff
@@ -0,0 +1,242 @@
+import io.airlift.units.Duration;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_CURRENT_STATE;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_MAX_WAIT;
+import static com.facebook.presto.spi.StandardErrorCode.REMOTE_TASK_MISMATCH;
+import static com.facebook.presto.util.Failures.REMOTE_TASK_MISMATCH_ERROR;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static com.google.common.net.MediaType.JSON_UTF_8;
+import static io.airlift.http.client.FullJsonResponseHandler.createFullJsonResponseHandler;
+import static io.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static io.airlift.http.client.Request.Builder.prepareGet;
+import static java.util.Objects.requireNonNull;
+
+class ContinuousTaskStatusFetcher
+        implements SimpleHttpResponseCallback<TaskStatus>
+{
+    private static final Logger log = Logger.get(ContinuousTaskStatusFetcher.class);
+
+    private final TaskId taskId;
+    private final HttpRemoteTask task;
+    private final StateMachine<TaskStatus> taskStatus;
+    private final JsonCodec<TaskStatus> taskStatusCodec;
+
+    private final Duration refreshMaxWait;
+    private final Executor executor;
+    private final HttpClient httpClient;
+    private final RequestErrorTracker errorTracker;
+
+    @GuardedBy("this")
+    private boolean running;
+
+    @GuardedBy("this")
+    private ListenableFuture<FullJsonResponseHandler.JsonResponse<TaskStatus>> future;
+
+    public ContinuousTaskStatusFetcher(
+            HttpRemoteTask task,
+            TaskStatus initialTaskStatus,
+            Duration refreshMaxWait,
+            JsonCodec<TaskStatus> taskStatusCodec,
+            Executor executor,
+            HttpClient httpClient,
+            Duration minErrorDuration,
+            ScheduledExecutorService errorScheduledExecutor)
+    {
+        requireNonNull(initialTaskStatus, "initialTaskStatus is null");
+
+        this.task = requireNonNull(task, "task is null");
+        this.taskId = initialTaskStatus.getTaskId();
+        this.taskStatus = new StateMachine<>("task " + taskId, executor, initialTaskStatus);
+
+        this.refreshMaxWait = requireNonNull(refreshMaxWait, "refreshMaxWait is null");
+        this.taskStatusCodec = requireNonNull(taskStatusCodec, "taskStatusCodec is null");
+
+        this.executor = requireNonNull(executor, "executor is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+
+        this.errorTracker = new RequestErrorTracker(taskId, initialTaskStatus.getSelf(), minErrorDuration, errorScheduledExecutor, "getting task status");
+    }
+
+    public synchronized void start()
```

Comment:
If start is always called immediately after construction, then you can make a static factory method that constructs and starts in one step.